File: 0874-walking-robot-simulation/0874-walking-robot-simulation.cpp
class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        // Change data structure
        unordered_set<string> lookup;
        for (vector<int>& obstacle : obstacles){
            lookup.insert(to_string(obstacle[0]) + " " + to_string(obstacle[1]));
        }
        
        // Directions: North, East, South, West
        vector<pair<int, int>> directions {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int currTurn = 0, maxDist = 0;
        pair<int, int> currCoordinates{0,0};
        
        for (int command : commands){
            if (command == -1){
                currTurn = (currTurn + 1) % 4;
                continue;
            }
            else if (command == -2){
                currTurn = (currTurn - 1 + 4) % 4;
                continue;
            }
            else{
                // If we'll meet the obstacles, then updated it to sign there's an obstacle.
                bool updated = false;
                
                for (int k = 1; k <= min(command, 9); ++k){
                    if (lookup.count(to_string(currCoordinates.first + directions[currTurn].first * k) + " " + to_string(currCoordinates.second + directions[currTurn].second * k))){
                        if (directions[currTurn].first == 1)
                            currCoordinates.first += directions[currTurn].first * k - 1;
                        else if (directions[currTurn].first == -1)
                            currCoordinates.first += directions[currTurn].first * k + 1;
                        else if (directions[currTurn].second == 1)
                            currCoordinates.second += directions[currTurn].second * k - 1;
                        else
                            currCoordinates.second += directions[currTurn].second * k + 1;
                        updated = true;
                        break;
                    }
                }
                
                if (!updated){
                    // Update the coordinates to move
                    currCoordinates.first += directions[currTurn].first * command;
                    currCoordinates.second += directions[currTurn].second * command;
                }
                
                maxDist = max(maxDist, currCoordinates.first * currCoordinates.first + currCoordinates.second * currCoordinates.second);
            }
        }
        
        return maxDist;
    }
};

File: 0874-walking-robot-simulation/0874-walking-robot-simulation.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        # change data structure of obstacles as set
        obstacles = set([tuple(obstacle) for obstacle in obstacles])
        
        # Directions: North, East, South, West
        directions, curr_turn, curr_coordinates = [(0, 1), (1, 0), (0, -1), (-1, 0)], 0, [0, 0]
        max_dist = 0
        
        for command in commands:
            if command == -1:
                curr_turn = (curr_turn + 1) % 4
                continue
            elif command == -2:
                curr_turn = (curr_turn - 1) % 4
                continue
            else:
                # If we'll meet the obstacles, then updated it to sign there's an obstacle.
                updated = False
                for k in range(1, min(command + 1, 10)):
                    if tuple([curr_coordinates[0] + directions[curr_turn][0] * k, \
                        curr_coordinates[1] + directions[curr_turn][1] * k]) in obstacles:
                        if directions[curr_turn][0] == 1:
                            curr_coordinates[0] += directions[curr_turn][0] * k - 1
                        elif directions[curr_turn][0] == -1:
                            curr_coordinates[0] += directions[curr_turn][0] * k + 1
                        elif directions[curr_turn][1] == 1:
                            curr_coordinates[1] += directions[curr_turn][1] * k - 1
                        else:
                            curr_coordinates[1] += directions[curr_turn][1] * k + 1
                        updated = True
                        break
                    
                if not updated:
                    # Update the coordinates to move
                    curr_coordinates[0] += directions[curr_turn][0] * command
                    curr_coordinates[1] += directions[curr_turn][1] * command
                
                max_dist = max(max_dist, curr_coordinates[0] ** 2 + curr_coordinates[1] ** 2)
            
        return max_dist